client/backoff/backoff.go

```diff
@@ -64,62 +64,51 @@ func (b *Backoff) Reset() {
 // It will backoff if the function returns a retryable error.
 // Once the context is done, retries will end and the most recent error will be returned.
 // Backoff is not reset by this function.
-func (b *Backoff) Retry(ctx context.Context, f func() error) error {
+func (b *Backoff) Retry(ctx context.Context, f func() error, retryableCodes ...codes.Code) error {
 	// If the context is already done, don't make any attempts to call f.
 	if ctx.Err() != nil {
 		return ctx.Err()
 	}
 
 	// Try calling f while the error is retryable and ctx is not done.
 	for {
-		if err := f(); IsRetryable(err) {
-			select {
-			case <-time.After(b.Duration()):
-				continue
-			case <-ctx.Done():
-				return err
-			}
+		if err := f(); !IsRetryable(err, retryableCodes...) {
+			return err
+		}
+		select {
+		case <-time.After(b.Duration()):
+		case <-ctx.Done():
+			return ctx.Err()
 		}
-		return nil
 	}
 }
 
-// IsRetryable returns true unless the error is explicitly not retriable per
-// https://godoc.org/google.golang.org/grpc/codes.
-func IsRetryable(err error) bool {
+// IsRetryable returns false unless the error is explicitly retriable per
+// https://godoc.org/google.golang.org/grpc/codes,
+// or if the error codes is in retry. codes.OK is not retryable.
+func IsRetryable(err error, retry ...codes.Code) bool {
 	switch code := status.Code(err); code {
-
-	// Errors that should not be retried:
-	case codes.OK, // Success = done.
-		codes.InvalidArgument,    // Retrying won't fix an invalid argument.
-		codes.PermissionDenied,   // Retrying won't fix a permission denied error.
-		codes.Unauthenticated,    // Retrying won't fix an unauthenticated error.
-		codes.FailedPrecondition, // Client must wait until the system has been fixed.
-		codes.Unimplemented,      // Retrying won't fix this.
-		codes.Internal,           // Something is very broken.
-		codes.DataLoss,           // Something is very broken.
-		codes.AlreadyExists,      // Retrying won't change this.
-		codes.Canceled:           // The client canceled: stop the retry loop.
-		return false
-
-	// Cases that are debatable:
-	case codes.OutOfRange, // Might be retryable if the server does a lot of work.
-		codes.NotFound: // Retrying might help, but not rapidly.
+	// Fast path.
+	case codes.OK:
 		return false
 
-	// Debatable cases. Retry to preserve existing behavior.
-	case codes.DeadlineExceeded, // The client or the server expired a timeout.
-		codes.ResourceExhausted, // Retry with backoff.
-		codes.Unknown:           // Catch-all error code.
+	// Debatable cases:
+	case codes.DeadlineExceeded,
+		codes.ResourceExhausted: // Retry with backoff.
 		return true
 
 	// Errors that are explicitly retryable:
 	case codes.Unavailable, // Client can just retry the call.
 		codes.Aborted: // Client can retry the read-modify-write function.
 		return true
 
-	// Retry for all unknown errors to preserve existing behavior.
 	default:
-		return true
+		for _, c := range retry {
```

Comment:
The previous code was a bit more flexible because it allowed to retry unretryable errors if you pass their codes in. Now they are unretryable forever.

Reply:
I did think about that, and as a result, I removed all cases of `return false` above so as to not fall into that trap.

Reply:
Ah, right, SGTM.

Also, optional nit: slightly a personal thing, but I would put the loop after the switch, and remove `default`, to have less indentation.